File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.cpp
#include <string>
#include <vector>

using namespace std;

int toInt(string ingredient)
{
    if (ingredient == "diamond")
        return 0;

    if (ingredient == "iron")
        return 1;

    if (ingredient == "stone")
        return 2;
}



int dfs(vector<int> picks, vector<vector<int>> fatigueList, int index, int fatigue) 
{
    // 곡괭이를 모두 썼거나, 이미 광물을 모두 캔 경우
    if (picks[0] + picks[1] + picks[2] == 0 || index >= fatigueList.size())
        return fatigue;

    int minFatigue = 749;

    for (int k = 0; k < 3; k++)
    {
        // k번째 곡괭이를 사용하여 광물을 캐는 경우
        if (picks[k] > 0)
        {
            picks[k]--;

            int tmp = dfs(picks, fatigueList, index + 1, fatigue + fatigueList[index][k]);

            if (minFatigue > tmp)
                minFatigue = tmp;

            picks[k]++;
        }
    }

    return minFatigue;

}


int solution(vector<int> picks, vector<string> minerals) {
    vector<int> intMinerals;

    for (auto mineral : minerals)
        intMinerals.push_back(toInt(mineral));

    vector<vector<int>> fatigueBoard = { {1,1,1}, {5,1,1}, {25,5,1} };

    vector<vector<int>> fatigueList;

    for (int i = 0; i < intMinerals.size(); i += 5)
    {
        vector<int> tmp(3, 0);

        int j = 0;
        while (j < 5 && i + j < intMinerals.size())
        {
            for(int k = 0; k < 3; k++)
                tmp[k] += fatigueBoard[k][intMinerals[i+j]];
            j++;
        }

        fatigueList.push_back(tmp);
    }

    return dfs(picks, fatigueList, 0, 0);
}

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.py
def solution(picks, minerals):

    remainPicks = picks
    groups = int(len(minerals) / 5)
    eachGroupSum = []
    sum1 = 0
    remains = len(minerals) % 5 

    sum2 = 0
    for x in picks:
        sum2 += x

    if remains != 0:
        groups += 1

    if groups - sum2 > 0 :
        minerals = minerals[0:sum2*5]

    for i in range(len(minerals)):
        if minerals[i] == 'diamond':
            sum1 += 25
        elif minerals[i] == 'iron':
            sum1 += 5
        elif minerals[i] == 'stone':
            sum1 += 1

        if i % 5 == 4 and i > 0:
            eachGroupSum.append(sum1)
            sum1 = 0

        if i % 5 != 4 and i == len(minerals) - 1 :
            eachGroupSum.append(sum1)

    indexed_list = list(enumerate(eachGroupSum))
    sorted_indexed_list = sorted(indexed_list, key=lambda x: x[1])
    original_indices = [index for index, value in sorted_indexed_list] # 원래 인덱스
    original_indices.reverse()
    print(original_indices)
    eachGroupSum.sort(reverse=True)

    fatigue = 0
    for i in range(len(eachGroupSum)):
        picks, pickaxe = getPickaxe(picks)
        print(picks, pickaxe)
        if pickaxe == 'none':
            break
        fatigue += getFatigue(original_indices[i], minerals, pickaxe)

    answer = fatigue
    return answer

def getPickaxe(picks):
    pick = ''
    if picks[0] > 0:
        picks[0] = picks[0] - 1
        pickaxe = 'diamondPA'
    elif picks[1] > 0:
        picks[1] = picks[1] - 1
        pickaxe = 'ironPA'
    elif picks[2] > 0:
        picks[2] = picks[2] - 1
        pickaxe = 'stonePA'
    else:
        pickaxe = 'none' 
    return picks, pickaxe

def getFatigue(i, minerals, pickaxe):
    fatigue = 0
    curMinerals = minerals[i*5:i*5+5]
    print(curMinerals)
    for j in range(len(curMinerals)):
        if curMinerals[j] == 'stone':
            fatigue += 1
        elif curMinerals[j] == 'iron':
            if pickaxe == 'stonePA':
                fatigue += 5
            else:
                fatigue += 1
        else:
            if pickaxe == 'stonePA':
                fatigue += 25
            elif pickaxe == 'ironPA':
                fatigue += 5
            else:
                fatigue += 1

    return fatigue   